src/constraint.rs

```diff
@@ -412,11 +424,136 @@ impl OrderedElementsConstraint {
             .collect::<IonSchemaResult<Vec<TypeId>>>()?;
         Ok(OrderedElementsConstraint::new(resolved_types))
     }
+
+    /// Returns an occurs constraint as range if it exists in the type_def otherwise returns `occurs: required`
+    fn get_occurs_constraint(type_def: &TypeDefinition) -> IonSchemaResult<Range> {
+        // verify if the type_def contains `occurs` constraint and fill occurs_range
+        // Otherwise if there is no `occurs` constraint specified then use `occurs: required`
+        if let Some(Constraint::Occurs(occurs)) = type_def
+            .constraints()
+            .iter()
+            .filter(|c| matches!(c, Constraint::Occurs(_)))
+            .next()
+        {
+            return occurs.isl_occurs().try_into();
+        }
+        // by default, if there is no `occurs` constraint for given type_def then use `occurs: required`
+        Range::required()
+    }
+
+    /// Validates a type_def for occurs constraint using values_iter
+    fn type_def_occurs_validation<'a>(
+        type_def: &TypeDefinition,
+        values_iter: &mut Peekable<Box<dyn Iterator<Item = &OwnedElement> + 'a>>,
+        type_store: &TypeStore,
+    ) -> ValidationResult {
+        let occurs_range: Range = OrderedElementsConstraint::get_occurs_constraint(type_def)
+            .expect("Unable to parse occurs to range");
```

Comment:
This currently panics. Can we propagate the error instead? (This can be a `TODO` if `ValidationResult` isn't designed for that.)

Reply:
I will create an issue for this.

Reply:
I didn't see an issue opened for it so I'm commenting here.

Based on the error message, it looks like the occurs range is being parsed in the validate method. The fundamental problem is not that the error is panicking instead of propagating; the problem is that parsing should already be done before we get around to validating. By the time we're validating, we should already know that the entire type definition (including the constraint arguments) is well-formed, and there should not be any possible error path when trying to get an occurs range from a type definition.